stdlib/marshal.gr

```diff
@@ -0,0 +1,1068 @@
+/**
+ * @module Marshal: Utilities for serializing and deserializing Grain data.
+ *
+ * @example import Marshal from "marshal"
+ *
+ * @since v0.5.3
+ */
+
+/*
+  SERIALIZED BINARY FORMAT
+
+  Grain stack-allocated values are serialized as-is. Heap-allocated values are
+  largely serialized the same as their in-memory representation, with the
+  exception that pointers are offsets into the byte sequence rather than
+  pointers into memory. Structures are kept 8-byte aligned to be consistent
+  with Grain's in-memory representation. The order in which these structures
+  appear in the byte sequence is not significant.
+
+  The first 32-bit value in the byte sequence is either a stack-allocated value
+  or a "pointer" to a heap-allocated value.
+*/
+
+/**
+ * @section Values: Functions for marshaling and unmarshaling data.
+ */
+
+import WasmI32, {
+  add as (+),
+  mul as (*),
+  and as (&),
+  eq as (==),
+  ne as (!=),
+  gtU as (>),
+  ltU as (<),
+  leU as (<=),
+  load,
+  store,
+  fromGrain,
+  toGrain,
+} from "runtime/unsafe/wasmi32"
+import WasmI64 from "runtime/unsafe/wasmi64"
+import Memory from "runtime/unsafe/memory"
+import Tags from "runtime/unsafe/tags"
+import { allocateBytes, newInt32 } from "runtime/dataStructures"
+import Map from "map"
+import Option from "option"
+
+@unsafe
+let roundTo8 = n => {
+  n + 7n & 0xfffffff8n
+}
+
+@unsafe
+let isHeapPtr = value =>
+  (value & Tags._GRAIN_GENERIC_TAG_MASK) ==
+  Tags._GRAIN_GENERIC_HEAP_TAG_TYPE
+
+@unsafe
+let rec size = (value, acc, valuesSeen, toplevel) => {
+  if (isHeapPtr(value)) {
+    let asInt32 = toGrain(newInt32(value)): Int32
+    if (Map.contains(asInt32, valuesSeen)) {
+      // We've detected a cycle, and we'll refer the existing instance
+      acc
+    } else {
+      Map.set(asInt32, void, valuesSeen)
```

Comment:
Why are you setting `void` here? Should this be a `Set`?

Reply:
Just to avoid depending on another library since we've already got maps imported. I was worried about having the unnecessary module size increase, but I checked and the difference is only 6k.